File: php/39-combination-sum.php
<?php
class Solution {
    private $target;
    private $result = [];
    private $count = 0;
    /**
     * @param Integer[] $candidates
     * @param Integer $target
     * @return Integer[][]
     */
    function combinationSum($candidates, $target) {
        if (empty($candidates)) {
            return [];
        }
        // sort and foreach break:Runtime: 12 ms, faster than 100.00%
        sort($candidates);
        $this->target = $target;
        $this->count = count($candidates);
        $this->helper($candidates, 0, []);
        return $this->result;
    }
    function helper(&$candidates, $index, $value) {
        $sum = array_sum($value);
        if ($sum > $this->target) {
            return;
        }
        if ($sum == $this->target) {
            $this->result[] = $value;
            return;
        }
        for ($i = $index; $i < $this->count; $i++) {
            if ($sum + $candidates[$i] > $this->target) {
                break;
            } 
            array_push($value, $candidates[$i]);
            $this->helper($candidates, $i, $value);
            array_pop($value);
        }
    }
}

File: go/39-combination-sum.go
package main

func combinationSum(candidates []int, target int) [][]int {
	res := [][]int{}
	if len(candidates) == 0 {
		return res
	}
	sort.Ints(candidates)
	start := 0
	helper(candidates, target, start, []int{}, &res)

	return res
}

func helper(candidates []int, target int, start int, el []int, res *[][]int) {
	if 0 == target {
		*res = append(*res, append([]int{}, el...))
		return
	}
	for i := start; i < len(candidates); i++ {
		if target-candidates[i] < 0 {
			continue
		}
		helper(candidates, target-candidates[i], i, append(el, candidates[i]), res)
	}
}
